// @strudel/mini/dist/index.mjs
import * as h from "@strudel/core";
function Ue(s, n) {
  function e() {
    this.constructor = s;
  }
  e.prototype = n.prototype, s.prototype = new e();
}
function N(s, n, e, c) {
  var l = Error.call(this, s);
  return Object.setPrototypeOf && Object.setPrototypeOf(l, N.prototype), l.expected = n, l.found = e, l.location = c, l.name = "SyntaxError", l;
}
Ue(N, Error);
function Au(s, n, e) {
  return e = e || " ", s.length > n ? s : (n -= s.length, e += e.repeat(n), s + e.slice(0, n));
}
N.prototype.format = function(s) {
  var n = "Error: " + this.message;
  if (this.location) {
    var e = null, c;
    for (c = 0; c < s.length; c++)
      if (s[c].source === this.location.source) {
        e = s[c].text.split(/\r\n|\n|\r/g);
        break;
      }
    var l = this.location.start, p = this.location.source && typeof this.location.source.offset == "function" ? this.location.source.offset(l) : l, m = this.location.source + ":" + p.line + ":" + p.column;
    if (e) {
      var $ = this.location.end, _ = Au("", p.line.toString().length, " "), i = e[l.line - 1], B = l.line === $.line ? $.column : i.length + 1, g = B - l.column || 1;
      n += `
 --> ` + m + `
` + _ + ` |
` + p.line + " | " + i + `
` + _ + " | " + Au("", l.column - 1, " ") + Au("", g, "^");
    } else
      n += `
 at ` + m;
  }
  return n;
};
N.buildMessage = function(s, n) {
  var e = {
    literal: function(i) {
      return '"' + l(i.text) + '"';
    },
    class: function(i) {
      var B = i.parts.map(function(g) {
        return Array.isArray(g) ? p(g[0]) + "-" + p(g[1]) : p(g);
      });
      return "[" + (i.inverted ? "^" : "") + B.join("") + "]";
    },
    any: function() {
      return "any character";
    },
    end: function() {
      return "end of input";
    },
    other: function(i) {
      return i.description;
    }
  };
  function c(i) {
    return i.charCodeAt(0).toString(16).toUpperCase();
  }
  function l(i) {
    return i.replace(/\\/g, "\\\\").replace(/"/g, '\\"').replace(/\0/g, "\\0").replace(/\t/g, "\\t").replace(/\n/g, "\\n").replace(/\r/g, "\\r").replace(/[\x00-\x0F]/g, function(B) {
      return "\\x0" + c(B);
    }).replace(/[\x10-\x1F\x7F-\x9F]/g, function(B) {
      return "\\x" + c(B);
    });
  }
  function p(i) {
    return i.replace(/\\/g, "\\\\").replace(/\]/g, "\\]").replace(/\^/g, "\\^").replace(/-/g, "\\-").replace(/\0/g, "\\0").replace(/\t/g, "\\t").replace(/\n/g, "\\n").replace(/\r/g, "\\r").replace(/[\x00-\x0F]/g, function(B) {
      return "\\x0" + c(B);
    }).replace(/[\x10-\x1F\x7F-\x9F]/g, function(B) {
      return "\\x" + c(B);
    });
  }
  function m(i) {
    return e[i.type](i);
  }
  function $(i) {
    var B = i.map(m), g, d;
    if (B.sort(), B.length > 0) {
      for (g = 1, d = 1; g < B.length; g++)
        B[g - 1] !== B[g] && (B[d] = B[g], d++);
      B.length = d;
    }
    switch (B.length) {
      case 1:
        return B[0];
      case 2:
        return B[0] + " or " + B[1];
      default:
        return B.slice(0, -1).join(", ") + ", or " + B[B.length - 1];
    }
  }
  function _(i) {
    return i ? '"' + l(i) + '"' : "end of input";
  }
  return "Expected " + $(s) + " but " + _(n) + " found.";
};
function Xe(s, n) {
  n = n !== void 0 ? n : {};
  var e = {}, c = n.grammarSource, l = { start: ku }, p = ku, m = ".", $ = "-", _ = "+", i = "0", B = ",", g = "|", d = '"', k = "'", z = "#", M = "^", ru = "_", Eu = "[", su = "]", Wu = "{", Zu = "}", Gu = "%", Uu = "<", Xu = ">", Hu = "@", Ju = "!", Ku = "(", Qu = ")", Vu = "/", Yu = "*", u0 = "?", e0 = ":", Cu = "..", Fu = "struct", tu = "target", nu = "euclid", au = "slow", fu = "rotL", iu = "rotR", Bu = "fast", Du = "scale", ou = "//", cu = "cat", A0 = "$", lu = "setcps", pu = "setbpm", gu = "hush", r0 = /^[1-9]/, E0 = /^[eE]/, s0 = /^[0-9]/, vu = /^[ \n\r\t\xA0]/, C0 = /^[0-9~]/, $u = /^[^\n]/, F0 = /^[a-z\xB5\xDF-\xF6\xF8-\xFF\u0101\u0103\u0105\u0107\u0109\u010B\u010D\u010F\u0111\u0113\u0115\u0117\u0119\u011B\u011D\u011F\u0121\u0123\u0125\u0127\u0129\u012B\u012D\u012F\u0131\u0133\u0135\u0137-\u0138\u013A\u013C\u013E\u0140\u0142\u0144\u0146\u0148-\u0149\u014B\u014D\u014F\u0151\u0153\u0155\u0157\u0159\u015B\u015D\u015F\u0161\u0163\u0165\u0167\u0169\u016B\u016D\u016F\u0171\u0173\u0175\u0177\u017A\u017C\u017E-\u0180\u0183\u0185\u0188\u018C-\u018D\u0192\u0195\u0199-\u019B\u019E\u01A1\u01A3\u01A5\u01A8\u01AA-\u01AB\u01AD\u01B0\u01B4\u01B6\u01B9-\u01BA\u01BD-\u01BF\u01C6\u01C9\u01CC\u01CE\u01D0\u01D2\u01D4\u01D6\u01D8\u01DA\u01DC-\u01DD\u01DF\u01E1\u01E3\u01E5\u01E7\u01E9\u01EB\u01ED\u01EF-\u01F0\u01F3\u01F5\u01F9\u01FB\u01FD\u01FF\u0201\u0203\u0205\u0207\u0209\u020B\u020D\u020F\u0211\u0213\u0215\u0217\u0219\u021B\u021D\u021F\u0221\u0223\u0225\u0227\u0229\u022B\u022D\u022F\u0231\u0233-\u0239\u023C\u023F-\u0240\u0242\u0247\u0249\u024B\u024D\u024F-\u0293\u0295-\u02AF\u0371\u0373\u0377\u037B-\u037D\u0390\u03AC-\u03CE\u03D0-\u03D1\u03D5-\u03D7\u03D9\u03DB\u03DD\u03DF\u03E1\u03E3\u03E5\u03E7\u03E9\u03EB\u03ED\u03EF-\u03F3\u03F5\u03F8\u03FB-\u03FC\u0430-\u045F\u0461\u0463\u0465\u0467\u0469\u046B\u046D\u046F\u0471\u0473\u0475\u0477\u0479\u047B\u047D\u047F\u0481\u048B\u048D\u048F\u0491\u0493\u0495\u0497\u0499\u049B\u049D\u049F\u04A1\u04A3\u04A5\u04A7\u04A9\u04AB\u04AD\u04AF\u04B1\u04B3\u04B5\u04B7\u04B9\u04BB\u04BD\u04BF\u04C2\u04C4\u04C6\u04C8\u04CA\u04CC\u04CE-\u04CF\u04D1\u04D3\u04D5\u04D7\u04D9\u04DB\u04DD\u04DF\u04E1\u04E3\u04E5\u04E7\u04E9\u04EB\u04ED\u04EF\u04F1\u04F3\u04F5\u04F7\u04F9\u04FB\u04FD\u04FF\u0501\u0503\u0505\u0507\u0509\u050B\u050D\u050F\u0511\u0513\u0515\u0517\u0519\u051B\u051D\u051F\u0521\u0523\u0525\u0527\u0529\u052B\u052D\u052F\u0560-\u0588\u10D0-\u10FA\u10FD-\u10FF\u13F8-\u13FD\u1C80-\u1C88\u1D00-\u1D2B\u1D6B-\u1D77\u1D79-\u1D9A\u1E01\u1E03\u1E05\u1E07\u1E09\u1E0B\u1E0D\u1E0F\u1E11\u1E13\u1E15\u1E17\u1E19\u1E1B\u1E1D\u1E1F\u1E21\u1E23\u1E25\u1E27\u1E29\u1E2B\u1E2D\u1E2F\u1E31\u1E33\u1E35\u1E37\u1E39\u1E3B\u1E3D\u1E3F\u1E41\u1E43\u1E45\u1E47\u1E49\u1E4B\u1E4D\u1E4F\u1E51\u1E53\u1E55\u1E57\u1E59\u1E5B\u1E5D\u1E5F\u1E61\u1E63\u1E65\u1E67\u1E69\u1E6B\u1E6D\u1E6F\u1E71\u1E73\u1E75\u1E77\u1E79\u1E7B\u1E7D\u1E7F\u1E81\u1E83\u1E85\u1E87\u1E89\u1E8B\u1E8D\u1E8F\u1E91\u1E93\u1E95-\u1E9D\u1E9F\u1EA1\u1EA3\u1EA5\u1EA7\u1EA9\u1EAB\u1EAD\u1EAF\u1EB1\u1EB3\u1EB5\u1EB7\u1EB9\u1EBB\u1EBD\u1EBF\u1EC1\u1EC3\u1EC5\u1EC7\u1EC9\u1ECB\u1ECD\u1ECF\u1ED1\u1ED3\u1ED5\u1ED7\u1ED9\u1EDB\u1EDD\u1EDF\u1EE1\u1EE3\u1EE5\u1EE7\u1EE9\u1EEB\u1EED\u1EEF\u1EF1\u1EF3\u1EF5\u1EF7\u1EF9\u1EFB\u1EFD\u1EFF-\u1F07\u1F10-\u1F15\u1F20-\u1F27\u1F30-\u1F37\u1F40-\u1F45\u1F50-\u1F57\u1F60-\u1F67\u1F70-\u1F7D\u1F80-\u1F87\u1F90-\u1F97\u1FA0-\u1FA7\u1FB0-\u1FB4\u1FB6-\u1FB7\u1FBE\u1FC2-\u1FC4\u1FC6-\u1FC7\u1FD0-\u1FD3\u1FD6-\u1FD7\u1FE0-\u1FE7\u1FF2-\u1FF4\u1FF6-\u1FF7\u210A\u210E-\u210F\u2113\u212F\u2134\u2139\u213C-\u213D\u2146-\u2149\u214E\u2184\u2C30-\u2C5E\u2C61\u2C65-\u2C66\u2C68\u2C6A\u2C6C\u2C71\u2C73-\u2C74\u2C76-\u2C7B\u2C81\u2C83\u2C85\u2C87\u2C89\u2C8B\u2C8D\u2C8F\u2C91\u2C93\u2C95\u2C97\u2C99\u2C9B\u2C9D\u2C9F\u2CA1\u2CA3\u2CA5\u2CA7\u2CA9\u2CAB\u2CAD\u2CAF\u2CB1\u2CB3\u2CB5\u2CB7\u2CB9\u2CBB\u2CBD\u2CBF\u2CC1\u2CC3\u2CC5\u2CC7\u2CC9\u2CCB\u2CCD\u2CCF\u2CD1\u2CD3\u2CD5\u2CD7\u2CD9\u2CDB\u2CDD\u2CDF\u2CE1\u2CE3-\u2CE4\u2CEC\u2CEE\u2CF3\u2D00-\u2D25\u2D27\u2D2D\uA641\uA643\uA645\uA647\uA649\uA64B\uA64D\uA64F\uA651\uA653\uA655\uA657\uA659\uA65B\uA65D\uA65F\uA661\uA663\uA665\uA667\uA669\uA66B\uA66D\uA681\uA683\uA685\uA687\uA689\uA68B\uA68D\uA68F\uA691\uA693\uA695\uA697\uA699\uA69B\uA723\uA725\uA727\uA729\uA72B\uA72D\uA72F-\uA731\uA733\uA735\uA737\uA739\uA73B\uA73D\uA73F\uA741\uA743\uA745\uA747\uA749\uA74B\uA74D\uA74F\uA751\uA753\uA755\uA757\uA759\uA75B\uA75D\uA75F\uA761\uA763\uA765\uA767\uA769\uA76B\uA76D\uA76F\uA771-\uA778\uA77A\uA77C\uA77F\uA781\uA783\uA785\uA787\uA78C\uA78E\uA791\uA793-\uA795\uA797\uA799\uA79B\uA79D\uA79F\uA7A1\uA7A3\uA7A5\uA7A7\uA7A9\uA7AF\uA7B5\uA7B7\uA7B9\uA7FA\uAB30-\uAB5A\uAB60-\uAB65\uAB70-\uABBF\uFB00-\uFB06\uFB13-\uFB17\uFF41-\uFF5A]/, t0 = /^[\u02B0-\u02C1\u02C6-\u02D1\u02E0-\u02E4\u02EC\u02EE\u0374\u037A\u0559\u0640\u06E5-\u06E6\u07F4-\u07F5\u07FA\u081A\u0824\u0828\u0971\u0E46\u0EC6\u10FC\u17D7\u1843\u1AA7\u1C78-\u1C7D\u1D2C-\u1D6A\u1D78\u1D9B-\u1DBF\u2071\u207F\u2090-\u209C\u2C7C-\u2C7D\u2D6F\u2E2F\u3005\u3031-\u3035\u303B\u309D-\u309E\u30FC-\u30FE\uA015\uA4F8-\uA4FD\uA60C\uA67F\uA69C-\uA69D\uA717-\uA71F\uA770\uA788\uA7F8-\uA7F9\uA9CF\uA9E6\uAA70\uAADD\uAAF3-\uAAF4\uAB5C-\uAB5F\uFF70\uFF9E-\uFF9F]/, n0 = /^[\xAA\xBA\u01BB\u01C0-\u01C3\u0294\u05D0-\u05EA\u05EF-\u05F2\u0620-\u063F\u0641-\u064A\u066E-\u066F\u0671-\u06D3\u06D5\u06EE-\u06EF\u06FA-\u06FC\u06FF\u0710\u0712-\u072F\u074D-\u07A5\u07B1\u07CA-\u07EA\u0800-\u0815\u0840-\u0858\u0860-\u086A\u08A0-\u08B4\u08B6-\u08BD\u0904-\u0939\u093D\u0950\u0958-\u0961\u0972-\u0980\u0985-\u098C\u098F-\u0990\u0993-\u09A8\u09AA-\u09B0\u09B2\u09B6-\u09B9\u09BD\u09CE\u09DC-\u09DD\u09DF-\u09E1\u09F0-\u09F1\u09FC\u0A05-\u0A0A\u0A0F-\u0A10\u0A13-\u0A28\u0A2A-\u0A30\u0A32-\u0A33\u0A35-\u0A36\u0A38-\u0A39\u0A59-\u0A5C\u0A5E\u0A72-\u0A74\u0A85-\u0A8D\u0A8F-\u0A91\u0A93-\u0AA8\u0AAA-\u0AB0\u0AB2-\u0AB3\u0AB5-\u0AB9\u0ABD\u0AD0\u0AE0-\u0AE1\u0AF9\u0B05-\u0B0C\u0B0F-\u0B10\u0B13-\u0B28\u0B2A-\u0B30\u0B32-\u0B33\u0B35-\u0B39\u0B3D\u0B5C-\u0B5D\u0B5F-\u0B61\u0B71\u0B83\u0B85-\u0B8A\u0B8E-\u0B90\u0B92-\u0B95\u0B99-\u0B9A\u0B9C\u0B9E-\u0B9F\u0BA3-\u0BA4\u0BA8-\u0BAA\u0BAE-\u0BB9\u0BD0\u0C05-\u0C0C\u0C0E-\u0C10\u0C12-\u0C28\u0C2A-\u0C39\u0C3D\u0C58-\u0C5A\u0C60-\u0C61\u0C80\u0C85-\u0C8C\u0C8E-\u0C90\u0C92-\u0CA8\u0CAA-\u0CB3\u0CB5-\u0CB9\u0CBD\u0CDE\u0CE0-\u0CE1\u0CF1-\u0CF2\u0D05-\u0D0C\u0D0E-\u0D10\u0D12-\u0D3A\u0D3D\u0D4E\u0D54-\u0D56\u0D5F-\u0D61\u0D7A-\u0D7F\u0D85-\u0D96\u0D9A-\u0DB1\u0DB3-\u0DBB\u0DBD\u0DC0-\u0DC6\u0E01-\u0E30\u0E32-\u0E33\u0E40-\u0E45\u0E81-\u0E82\u0E84\u0E87-\u0E88\u0E8A\u0E8D\u0E94-\u0E97\u0E99-\u0E9F\u0EA1-\u0EA3\u0EA5\u0EA7\u0EAA-\u0EAB\u0EAD-\u0EB0\u0EB2-\u0EB3\u0EBD\u0EC0-\u0EC4\u0EDC-\u0EDF\u0F00\u0F40-\u0F47\u0F49-\u0F6C\u0F88-\u0F8C\u1000-\u102A\u103F\u1050-\u1055\u105A-\u105D\u1061\u1065-\u1066\u106E-\u1070\u1075-\u1081\u108E\u1100-\u1248\u124A-\u124D\u1250-\u1256\u1258\u125A-\u125D\u1260-\u1288\u128A-\u128D\u1290-\u12B0\u12B2-\u12B5\u12B8-\u12BE\u12C0\u12C2-\u12C5\u12C8-\u12D6\u12D8-\u1310\u1312-\u1315\u1318-\u135A\u1380-\u138F\u1401-\u166C\u166F-\u167F\u1681-\u169A\u16A0-\u16EA\u16F1-\u16F8\u1700-\u170C\u170E-\u1711\u1720-\u1731\u1740-\u1751\u1760-\u176C\u176E-\u1770\u1780-\u17B3\u17DC\u1820-\u1842\u1844-\u1878\u1880-\u1884\u1887-\u18A8\u18AA\u18B0-\u18F5\u1900-\u191E\u1950-\u196D\u1970-\u1974\u1980-\u19AB\u19B0-\u19C9\u1A00-\u1A16\u1A20-\u1A54\u1B05-\u1B33\u1B45-\u1B4B\u1B83-\u1BA0\u1BAE-\u1BAF\u1BBA-\u1BE5\u1C00-\u1C23\u1C4D-\u1C4F\u1C5A-\u1C77\u1CE9-\u1CEC\u1CEE-\u1CF1\u1CF5-\u1CF6\u2135-\u2138\u2D30-\u2D67\u2D80-\u2D96\u2DA0-\u2DA6\u2DA8-\u2DAE\u2DB0-\u2DB6\u2DB8-\u2DBE\u2DC0-\u2DC6\u2DC8-\u2DCE\u2DD0-\u2DD6\u2DD8-\u2DDE\u3006\u303C\u3041-\u3096\u309F\u30A1-\u30FA\u30FF\u3105-\u312F\u3131-\u318E\u31A0-\u31BA\u31F0-\u31FF\u3400-\u4DB5\u4E00-\u9FEF\uA000-\uA014\uA016-\uA48C\uA4D0-\uA4F7\uA500-\uA60B\uA610-\uA61F\uA62A-\uA62B\uA66E\uA6A0-\uA6E5\uA78F\uA7F7\uA7FB-\uA801\uA803-\uA805\uA807-\uA80A\uA80C-\uA822\uA840-\uA873\uA882-\uA8B3\uA8F2-\uA8F7\uA8FB\uA8FD-\uA8FE\uA90A-\uA925\uA930-\uA946\uA960-\uA97C\uA984-\uA9B2\uA9E0-\uA9E4\uA9E7-\uA9EF\uA9FA-\uA9FE\uAA00-\uAA28\uAA40-\uAA42\uAA44-\uAA4B\uAA60-\uAA6F\uAA71-\uAA76\uAA7A\uAA7E-\uAAAF\uAAB1\uAAB5-\uAAB6\uAAB9-\uAABD\uAAC0\uAAC2\uAADB-\uAADC\uAAE0-\uAAEA\uAAF2\uAB01-\uAB06\uAB09-\uAB0E\uAB11-\uAB16\uAB20-\uAB26\uAB28-\uAB2E\uABC0-\uABE2\uAC00-\uD7A3\uD7B0-\uD7C6\uD7CB-\uD7FB\uF900-\uFA6D\uFA70-\uFAD9\uFB1D\uFB1F-\uFB28\uFB2A-\uFB36\uFB38-\uFB3C\uFB3E\uFB40-\uFB41\uFB43-\uFB44\uFB46-\uFBB1\uFBD3-\uFD3D\uFD50-\uFD8F\uFD92-\uFDC7\uFDF0-\uFDFB\uFE70-\uFE74\uFE76-\uFEFC\uFF66-\uFF6F\uFF71-\uFF9D\uFFA0-\uFFBE\uFFC2-\uFFC7\uFFCA-\uFFCF\uFFD2-\uFFD7\uFFDA-\uFFDC]/, a0 = /^[\u01C5\u01C8\u01CB\u01F2\u1F88-\u1F8F\u1F98-\u1F9F\u1FA8-\u1FAF\u1FBC\u1FCC\u1FFC]/, f0 = /^[A-Z\xC0-\xD6\xD8-\xDE\u0100\u0102\u0104\u0106\u0108\u010A\u010C\u010E\u0110\u0112\u0114\u0116\u0118\u011A\u011C\u011E\u0120\u0122\u0124\u0126\u0128\u012A\u012C\u012E\u0130\u0132\u0134\u0136\u0139\u013B\u013D\u013F\u0141\u0143\u0145\u0147\u014A\u014C\u014E\u0150\u0152\u0154\u0156\u0158\u015A\u015C\u015E\u0160\u0162\u0164\u0166\u0168\u016A\u016C\u016E\u0170\u0172\u0174\u0176\u0178-\u0179\u017B\u017D\u0181-\u0182\u0184\u0186-\u0187\u0189-\u018B\u018E-\u0191\u0193-\u0194\u0196-\u0198\u019C-\u019D\u019F-\u01A0\u01A2\u01A4\u01A6-\u01A7\u01A9\u01AC\u01AE-\u01AF\u01B1-\u01B3\u01B5\u01B7-\u01B8\u01BC\u01C4\u01C7\u01CA\u01CD\u01CF\u01D1\u01D3\u01D5\u01D7\u01D9\u01DB\u01DE\u01E0\u01E2\u01E4\u01E6\u01E8\u01EA\u01EC\u01EE\u01F1\u01F4\u01F6-\u01F8\u01FA\u01FC\u01FE\u0200\u0202\u0204\u0206\u0208\u020A\u020C\u020E\u0210\u0212\u0214\u0216\u0218\u021A\u021C\u021E\u0220\u0222\u0224\u0226\u0228\u022A\u022C\u022E\u0230\u0232\u023A-\u023B\u023D-\u023E\u0241\u0243-\u0246\u0248\u024A\u024C\u024E\u0370\u0372\u0376\u037F\u0386\u0388-\u038A\u038C\u038E-\u038F\u0391-\u03A1\u03A3-\u03AB\u03CF\u03D2-\u03D4\u03D8\u03DA\u03DC\u03DE\u03E0\u03E2\u03E4\u03E6\u03E8\u03EA\u03EC\u03EE\u03F4\u03F7\u03F9-\u03FA\u03FD-\u042F\u0460\u0462\u0464\u0466\u0468\u046A\u046C\u046E\u0470\u0472\u0474\u0476\u0478\u047A\u047C\u047E\u0480\u048A\u048C\u048E\u0490\u0492\u0494\u0496\u0498\u049A\u049C\u049E\u04A0\u04A2\u04A4\u04A6\u04A8\u04AA\u04AC\u04AE\u04B0\u04B2\u04B4\u04B6\u04B8\u04BA\u04BC\u04BE\u04C0-\u04C1\u04C3\u04C5\u04C7\u04C9\u04CB\u04CD\u04D0\u04D2\u04D4\u04D6\u04D8\u04DA\u04DC\u04DE\u04E0\u04E2\u04E4\u04E6\u04E8\u04EA\u04EC\u04EE\u04F0\u04F2\u04F4\u04F6\u04F8\u04FA\u04FC\u04FE\u0500\u0502\u0504\u0506\u0508\u050A\u050C\u050E\u0510\u0512\u0514\u0516\u0518\u051A\u051C\u051E\u0520\u0522\u0524\u0526\u0528\u052A\u052C\u052E\u0531-\u0556\u10A0-\u10C5\u10C7\u10CD\u13A0-\u13F5\u1C90-\u1CBA\u1CBD-\u1CBF\u1E00\u1E02\u1E04\u1E06\u1E08\u1E0A\u1E0C\u1E0E\u1E10\u1E12\u1E14\u1E16\u1E18\u1E1A\u1E1C\u1E1E\u1E20\u1E22\u1E24\u1E26\u1E28\u1E2A\u1E2C\u1E2E\u1E30\u1E32\u1E34\u1E36\u1E38\u1E3A\u1E3C\u1E3E\u1E40\u1E42\u1E44\u1E46\u1E48\u1E4A\u1E4C\u1E4E\u1E50\u1E52\u1E54\u1E56\u1E58\u1E5A\u1E5C\u1E5E\u1E60\u1E62\u1E64\u1E66\u1E68\u1E6A\u1E6C\u1E6E\u1E70\u1E72\u1E74\u1E76\u1E78\u1E7A\u1E7C\u1E7E\u1E80\u1E82\u1E84\u1E86\u1E88\u1E8A\u1E8C\u1E8E\u1E90\u1E92\u1E94\u1E9E\u1EA0\u1EA2\u1EA4\u1EA6\u1EA8\u1EAA\u1EAC\u1EAE\u1EB0\u1EB2\u1EB4\u1EB6\u1EB8\u1EBA\u1EBC\u1EBE\u1EC0\u1EC2\u1EC4\u1EC6\u1EC8\u1ECA\u1ECC\u1ECE\u1ED0\u1ED2\u1ED4\u1ED6\u1ED8\u1EDA\u1EDC\u1EDE\u1EE0\u1EE2\u1EE4\u1EE6\u1EE8\u1EEA\u1EEC\u1EEE\u1EF0\u1EF2\u1EF4\u1EF6\u1EF8\u1EFA\u1EFC\u1EFE\u1F08-\u1F0F\u1F18-\u1F1D\u1F28-\u1F2F\u1F38-\u1F3F\u1F48-\u1F4D\u1F59\u1F5B\u1F5D\u1F5F\u1F68-\u1F6F\u1FB8-\u1FBB\u1FC8-\u1FCB\u1FD8-\u1FDB\u1FE8-\u1FEC\u1FF8-\u1FFB\u2102\u2107\u210B-\u210D\u2110-\u2112\u2115\u2119-\u211D\u2124\u2126\u2128\u212A-\u212D\u2130-\u2133\u213E-\u213F\u2145\u2183\u2C00-\u2C2E\u2C60\u2C62-\u2C64\u2C67\u2C69\u2C6B\u2C6D-\u2C70\u2C72\u2C75\u2C7E-\u2C80\u2C82\u2C84\u2C86\u2C88\u2C8A\u2C8C\u2C8E\u2C90\u2C92\u2C94\u2C96\u2C98\u2C9A\u2C9C\u2C9E\u2CA0\u2CA2\u2CA4\u2CA6\u2CA8\u2CAA\u2CAC\u2CAE\u2CB0\u2CB2\u2CB4\u2CB6\u2CB8\u2CBA\u2CBC\u2CBE\u2CC0\u2CC2\u2CC4\u2CC6\u2CC8\u2CCA\u2CCC\u2CCE\u2CD0\u2CD2\u2CD4\u2CD6\u2CD8\u2CDA\u2CDC\u2CDE\u2CE0\u2CE2\u2CEB\u2CED\u2CF2\uA640\uA642\uA644\uA646\uA648\uA64A\uA64C\uA64E\uA650\uA652\uA654\uA656\uA658\uA65A\uA65C\uA65E\uA660\uA662\uA664\uA666\uA668\uA66A\uA66C\uA680\uA682\uA684\uA686\uA688\uA68A\uA68C\uA68E\uA690\uA692\uA694\uA696\uA698\uA69A\uA722\uA724\uA726\uA728\uA72A\uA72C\uA72E\uA732\uA734\uA736\uA738\uA73A\uA73C\uA73E\uA740\uA742\uA744\uA746\uA748\uA74A\uA74C\uA74E\uA750\uA752\uA754\uA756\uA758\uA75A\uA75C\uA75E\uA760\uA762\uA764\uA766\uA768\uA76A\uA76C\uA76E\uA779\uA77B\uA77D-\uA77E\uA780\uA782\uA784\uA786\uA78B\uA78D\uA790\uA792\uA796\uA798\uA79A\uA79C\uA79E\uA7A0\uA7A2\uA7A4\uA7A6\uA7A8\uA7AA-\uA7AE\uA7B0-\uA7B4\uA7B6\uA7B8\uFF21-\uFF3A]/, i0 = /^[\u16EE-\u16F0\u2160-\u2182\u2185-\u2188\u3007\u3021-\u3029\u3038-\u303A\uA6E6-\uA6EF]/, B0 = Q("number"), J = o(".", !1), D0 = y([["1", "9"]], !1, !1), o0 = y(["e", "E"], !1, !1), hu = o("-", !1), c0 = o("+", !1), l0 = o("0", !1), p0 = y([["0", "9"]], !1, !1), g0 = Q("whitespace"), mu = y([" ", `
`, "\r", "	", " "], !1, !1), v0 = o(",", !1), $0 = o("|", !1), h0 = o('"', !1), m0 = o("'", !1), _0 = Q('a letter, a number, "-", "#", ".", "^", "_"'), d0 = y([["0", "9"], "~"], !1, !1), y0 = o("#", !1), w0 = o("^", !1), _u = o("_", !1), du = o("[", !1), yu = o("]", !1), x0 = o("{", !1), b0 = o("}", !1), j0 = o("%", !1), k0 = o("<", !1), S0 = o(">", !1), L0 = o("@", !1), P0 = o("!", !1), R0 = o("(", !1), q0 = o(")", !1), I0 = o("/", !1), N0 = o("*", !1), O0 = o("?", !1), T0 = o(":", !1), z0 = o("..", !1), M0 = o("struct", !1), W0 = o("target", !1), Z0 = o("euclid", !1), G0 = o("slow", !1), U0 = o("rotL", !1), X0 = o("rotR", !1), H0 = o("fast", !1), J0 = o("scale", !1), K0 = o("//", !1), wu = y([`
`], !0, !1), Q0 = o("cat", !1), V0 = o("$", !1), Y0 = o("setcps", !1), u1 = o("setbpm", !1), e1 = o("hush", !1), A1 = y([["a", "z"], "µ", ["ß", "ö"], ["ø", "ÿ"], "ā", "ă", "ą", "ć", "ĉ", "ċ", "č", "ď", "đ", "ē", "ĕ", "ė", "ę", "ě", "ĝ", "ğ", "ġ", "ģ", "ĥ", "ħ", "ĩ", "ī", "ĭ", "į", "ı", "ĳ", "ĵ", ["ķ", "ĸ"], "ĺ", "ļ", "ľ", "ŀ", "ł", "ń", "ņ", ["ň", "ŉ"], "ŋ", "ō", "ŏ", "ő", "œ", "ŕ", "ŗ", "ř", "ś", "ŝ", "ş", "š", "ţ", "ť", "ŧ", "ũ", "ū", "ŭ", "ů", "ű", "ų", "ŵ", "ŷ", "ź", "ż", ["ž", "ƀ"], "ƃ", "ƅ", "ƈ", ["ƌ", "ƍ"], "ƒ", "ƕ", ["ƙ", "ƛ"], "ƞ", "ơ", "ƣ", "ƥ", "ƨ", ["ƪ", "ƫ"], "ƭ", "ư", "ƴ", "ƶ", ["ƹ", "ƺ"], ["ƽ", "ƿ"], "ǆ", "ǉ", "ǌ", "ǎ", "ǐ", "ǒ", "ǔ", "ǖ", "ǘ", "ǚ", ["ǜ", "ǝ"], "ǟ", "ǡ", "ǣ", "ǥ", "ǧ", "ǩ", "ǫ", "ǭ", ["ǯ", "ǰ"], "ǳ", "ǵ", "ǹ", "ǻ", "ǽ", "ǿ", "ȁ", "ȃ", "ȅ", "ȇ", "ȉ", "ȋ", "ȍ", "ȏ", "ȑ", "ȓ", "ȕ", "ȗ", "ș", "ț", "ȝ", "ȟ", "ȡ", "ȣ", "ȥ", "ȧ", "ȩ", "ȫ", "ȭ", "ȯ", "ȱ", ["ȳ", "ȹ"], "ȼ", ["ȿ", "ɀ"], "ɂ", "ɇ", "ɉ", "ɋ", "ɍ", ["ɏ", "ʓ"], ["ʕ", "ʯ"], "ͱ", "ͳ", "ͷ", ["ͻ", "ͽ"], "ΐ", ["ά", "ώ"], ["ϐ", "ϑ"], ["ϕ", "ϗ"], "ϙ", "ϛ", "ϝ", "ϟ", "ϡ", "ϣ", "ϥ", "ϧ", "ϩ", "ϫ", "ϭ", ["ϯ", "ϳ"], "ϵ", "ϸ", ["ϻ", "ϼ"], ["а", "џ"], "ѡ", "ѣ", "ѥ", "ѧ", "ѩ", "ѫ", "ѭ", "ѯ", "ѱ", "ѳ", "ѵ", "ѷ", "ѹ", "ѻ", "ѽ", "ѿ", "ҁ", "ҋ", "ҍ", "ҏ", "ґ", "ғ", "ҕ", "җ", "ҙ", "қ", "ҝ", "ҟ", "ҡ", "ң", "ҥ", "ҧ", "ҩ", "ҫ", "ҭ", "ү", "ұ", "ҳ", "ҵ", "ҷ", "ҹ", "һ", "ҽ", "ҿ", "ӂ", "ӄ", "ӆ", "ӈ", "ӊ", "ӌ", ["ӎ", "ӏ"], "ӑ", "ӓ", "ӕ", "ӗ", "ә", "ӛ", "ӝ", "ӟ", "ӡ", "ӣ", "ӥ", "ӧ", "ө", "ӫ", "ӭ", "ӯ", "ӱ", "ӳ", "ӵ", "ӷ", "ӹ", "ӻ", "ӽ", "ӿ", "ԁ", "ԃ", "ԅ", "ԇ", "ԉ", "ԋ", "ԍ", "ԏ", "ԑ", "ԓ", "ԕ", "ԗ", "ԙ", "ԛ", "ԝ", "ԟ", "ԡ", "ԣ", "ԥ", "ԧ", "ԩ", "ԫ", "ԭ", "ԯ", ["ՠ", "ֈ"], ["ა", "ჺ"], ["ჽ", "ჿ"], ["ᏸ", "ᏽ"], ["ᲀ", "ᲈ"], ["ᴀ", "ᴫ"], ["ᵫ", "ᵷ"], ["ᵹ", "ᶚ"], "ḁ", "ḃ", "ḅ", "ḇ", "ḉ", "ḋ", "ḍ", "ḏ", "ḑ", "ḓ", "ḕ", "ḗ", "ḙ", "ḛ", "ḝ", "ḟ", "ḡ", "ḣ", "ḥ", "ḧ", "ḩ", "ḫ", "ḭ", "ḯ", "ḱ", "ḳ", "ḵ", "ḷ", "ḹ", "ḻ", "ḽ", "ḿ", "ṁ", "ṃ", "ṅ", "ṇ", "ṉ", "ṋ", "ṍ", "ṏ", "ṑ", "ṓ", "ṕ", "ṗ", "ṙ", "ṛ", "ṝ", "ṟ", "ṡ", "ṣ", "ṥ", "ṧ", "ṩ", "ṫ", "ṭ", "ṯ", "ṱ", "ṳ", "ṵ", "ṷ", "ṹ", "ṻ", "ṽ", "ṿ", "ẁ", "ẃ", "ẅ", "ẇ", "ẉ", "ẋ", "ẍ", "ẏ", "ẑ", "ẓ", ["ẕ", "ẝ"], "ẟ", "ạ", "ả", "ấ", "ầ", "ẩ", "ẫ", "ậ", "ắ", "ằ", "ẳ", "ẵ", "ặ", "ẹ", "ẻ", "ẽ", "ế", "ề", "ể", "ễ", "ệ", "ỉ", "ị", "ọ", "ỏ", "ố", "ồ", "ổ", "ỗ", "ộ", "ớ", "ờ", "ở", "ỡ", "ợ", "ụ", "ủ", "ứ", "ừ", "ử", "ữ", "ự", "ỳ", "ỵ", "ỷ", "ỹ", "ỻ", "ỽ", ["ỿ", "ἇ"], ["ἐ", "ἕ"], ["ἠ", "ἧ"], ["ἰ", "ἷ"], ["ὀ", "ὅ"], ["ὐ", "ὗ"], ["ὠ", "ὧ"], ["ὰ", "ώ"], ["ᾀ", "ᾇ"], ["ᾐ", "ᾗ"], ["ᾠ", "ᾧ"], ["ᾰ", "ᾴ"], ["ᾶ", "ᾷ"], "ι", ["ῂ", "ῄ"], ["ῆ", "ῇ"], ["ῐ", "ΐ"], ["ῖ", "ῗ"], ["ῠ", "ῧ"], ["ῲ", "ῴ"], ["ῶ", "ῷ"], "ℊ", ["ℎ", "ℏ"], "ℓ", "ℯ", "ℴ", "ℹ", ["ℼ", "ℽ"], ["ⅆ", "ⅉ"], "ⅎ", "ↄ", ["ⰰ", "ⱞ"], "ⱡ", ["ⱥ", "ⱦ"], "ⱨ", "ⱪ", "ⱬ", "ⱱ", ["ⱳ", "ⱴ"], ["ⱶ", "ⱻ"], "ⲁ", "ⲃ", "ⲅ", "ⲇ", "ⲉ", "ⲋ", "ⲍ", "ⲏ", "ⲑ", "ⲓ", "ⲕ", "ⲗ", "ⲙ", "ⲛ", "ⲝ", "ⲟ", "ⲡ", "ⲣ", "ⲥ", "ⲧ", "ⲩ", "ⲫ", "ⲭ", "ⲯ", "ⲱ", "ⲳ", "ⲵ", "ⲷ", "ⲹ", "ⲻ", "ⲽ", "ⲿ", "ⳁ", "ⳃ", "ⳅ", "ⳇ", "ⳉ", "ⳋ", "ⳍ", "ⳏ", "ⳑ", "ⳓ", "ⳕ", "ⳗ", "ⳙ", "ⳛ", "ⳝ", "ⳟ", "ⳡ", ["ⳣ", "ⳤ"], "ⳬ", "ⳮ", "ⳳ", ["ⴀ", "ⴥ"], "ⴧ", "ⴭ", "ꙁ", "ꙃ", "ꙅ", "ꙇ", "ꙉ", "ꙋ", "ꙍ", "ꙏ", "ꙑ", "ꙓ", "ꙕ", "ꙗ", "ꙙ", "ꙛ", "ꙝ", "ꙟ", "ꙡ", "ꙣ", "ꙥ", "ꙧ", "ꙩ", "ꙫ", "ꙭ", "ꚁ", "ꚃ", "ꚅ", "ꚇ", "ꚉ", "ꚋ", "ꚍ", "ꚏ", "ꚑ", "ꚓ", "ꚕ", "ꚗ", "ꚙ", "ꚛ", "ꜣ", "ꜥ", "ꜧ", "ꜩ", "ꜫ", "ꜭ", ["ꜯ", "ꜱ"], "ꜳ", "ꜵ", "ꜷ", "ꜹ", "ꜻ", "ꜽ", "ꜿ", "ꝁ", "ꝃ", "ꝅ", "ꝇ", "ꝉ", "ꝋ", "ꝍ", "ꝏ", "ꝑ", "ꝓ", "ꝕ", "ꝗ", "ꝙ", "ꝛ", "ꝝ", "ꝟ", "ꝡ", "ꝣ", "ꝥ", "ꝧ", "ꝩ", "ꝫ", "ꝭ", "ꝯ", ["ꝱ", "ꝸ"], "ꝺ", "ꝼ", "ꝿ", "ꞁ", "ꞃ", "ꞅ", "ꞇ", "ꞌ", "ꞎ", "ꞑ", ["ꞓ", "ꞕ"], "ꞗ", "ꞙ", "ꞛ", "ꞝ", "ꞟ", "ꞡ", "ꞣ", "ꞥ", "ꞧ", "ꞩ", "ꞯ", "ꞵ", "ꞷ", "ꞹ", "ꟺ", ["ꬰ", "ꭚ"], ["ꭠ", "ꭥ"], ["ꭰ", "ꮿ"], ["ﬀ", "ﬆ"], ["ﬓ", "ﬗ"], ["ａ", "ｚ"]], !1, !1), r1 = y([["ʰ", "ˁ"], ["ˆ", "ˑ"], ["ˠ", "ˤ"], "ˬ", "ˮ", "ʹ", "ͺ", "ՙ", "ـ", ["ۥ", "ۦ"], ["ߴ", "ߵ"], "ߺ", "ࠚ", "ࠤ", "ࠨ", "ॱ", "ๆ", "ໆ", "ჼ", "ៗ", "ᡃ", "ᪧ", ["ᱸ", "ᱽ"], ["ᴬ", "ᵪ"], "ᵸ", ["ᶛ", "ᶿ"], "ⁱ", "ⁿ", ["ₐ", "ₜ"], ["ⱼ", "ⱽ"], "ⵯ", "ⸯ", "々", ["〱", "〵"], "〻", ["ゝ", "ゞ"], ["ー", "ヾ"], "ꀕ", ["ꓸ", "ꓽ"], "ꘌ", "ꙿ", ["ꚜ", "ꚝ"], ["ꜗ", "ꜟ"], "ꝰ", "ꞈ", ["ꟸ", "ꟹ"], "ꧏ", "ꧦ", "ꩰ", "ꫝ", ["ꫳ", "ꫴ"], ["ꭜ", "ꭟ"], "ｰ", ["ﾞ", "ﾟ"]], !1, !1), E1 = y(["ª", "º", "ƻ", ["ǀ", "ǃ"], "ʔ", ["א", "ת"], ["ׯ", "ײ"], ["ؠ", "ؿ"], ["ف", "ي"], ["ٮ", "ٯ"], ["ٱ", "ۓ"], "ە", ["ۮ", "ۯ"], ["ۺ", "ۼ"], "ۿ", "ܐ", ["ܒ", "ܯ"], ["ݍ", "ޥ"], "ޱ", ["ߊ", "ߪ"], ["ࠀ", "ࠕ"], ["ࡀ", "ࡘ"], ["ࡠ", "ࡪ"], ["ࢠ", "ࢴ"], ["ࢶ", "ࢽ"], ["ऄ", "ह"], "ऽ", "ॐ", ["क़", "ॡ"], ["ॲ", "ঀ"], ["অ", "ঌ"], ["এ", "ঐ"], ["ও", "ন"], ["প", "র"], "ল", ["শ", "হ"], "ঽ", "ৎ", ["ড়", "ঢ়"], ["য়", "ৡ"], ["ৰ", "ৱ"], "ৼ", ["ਅ", "ਊ"], ["ਏ", "ਐ"], ["ਓ", "ਨ"], ["ਪ", "ਰ"], ["ਲ", "ਲ਼"], ["ਵ", "ਸ਼"], ["ਸ", "ਹ"], ["ਖ਼", "ੜ"], "ਫ਼", ["ੲ", "ੴ"], ["અ", "ઍ"], ["એ", "ઑ"], ["ઓ", "ન"], ["પ", "ર"], ["લ", "ળ"], ["વ", "હ"], "ઽ", "ૐ", ["ૠ", "ૡ"], "ૹ", ["ଅ", "ଌ"], ["ଏ", "ଐ"], ["ଓ", "ନ"], ["ପ", "ର"], ["ଲ", "ଳ"], ["ଵ", "ହ"], "ଽ", ["ଡ଼", "ଢ଼"], ["ୟ", "ୡ"], "ୱ", "ஃ", ["அ", "ஊ"], ["எ", "ஐ"], ["ஒ", "க"], ["ங", "ச"], "ஜ", ["ஞ", "ட"], ["ண", "த"], ["ந", "ப"], ["ம", "ஹ"], "ௐ", ["అ", "ఌ"], ["ఎ", "ఐ"], ["ఒ", "న"], ["ప", "హ"], "ఽ", ["ౘ", "ౚ"], ["ౠ", "ౡ"], "ಀ", ["ಅ", "ಌ"], ["ಎ", "ಐ"], ["ಒ", "ನ"], ["ಪ", "ಳ"], ["ವ", "ಹ"], "ಽ", "ೞ", ["ೠ", "ೡ"], ["ೱ", "ೲ"], ["അ", "ഌ"], ["എ", "ഐ"], ["ഒ", "ഺ"], "ഽ", "ൎ", ["ൔ", "ൖ"], ["ൟ", "ൡ"], ["ൺ", "ൿ"], ["අ", "ඖ"], ["ක", "න"], ["ඳ", "ර"], "ල", ["ව", "ෆ"], ["ก", "ะ"], ["า", "ำ"], ["เ", "ๅ"], ["ກ", "ຂ"], "ຄ", ["ງ", "ຈ"], "ຊ", "ຍ", ["ດ", "ທ"], ["ນ", "ຟ"], ["ມ", "ຣ"], "ລ", "ວ", ["ສ", "ຫ"], ["ອ", "ະ"], ["າ", "ຳ"], "ຽ", ["ເ", "ໄ"], ["ໜ", "ໟ"], "ༀ", ["ཀ", "ཇ"], ["ཉ", "ཬ"], ["ྈ", "ྌ"], ["က", "ဪ"], "ဿ", ["ၐ", "ၕ"], ["ၚ", "ၝ"], "ၡ", ["ၥ", "ၦ"], ["ၮ", "ၰ"], ["ၵ", "ႁ"], "ႎ", ["ᄀ", "ቈ"], ["ቊ", "ቍ"], ["ቐ", "ቖ"], "ቘ", ["ቚ", "ቝ"], ["በ", "ኈ"], ["ኊ", "ኍ"], ["ነ", "ኰ"], ["ኲ", "ኵ"], ["ኸ", "ኾ"], "ዀ", ["ዂ", "ዅ"], ["ወ", "ዖ"], ["ዘ", "ጐ"], ["ጒ", "ጕ"], ["ጘ", "ፚ"], ["ᎀ", "ᎏ"], ["ᐁ", "ᙬ"], ["ᙯ", "ᙿ"], ["ᚁ", "ᚚ"], ["ᚠ", "ᛪ"], ["ᛱ", "ᛸ"], ["ᜀ", "ᜌ"], ["ᜎ", "ᜑ"], ["ᜠ", "ᜱ"], ["ᝀ", "ᝑ"], ["ᝠ", "ᝬ"], ["ᝮ", "ᝰ"], ["ក", "ឳ"], "ៜ", ["ᠠ", "ᡂ"], ["ᡄ", "ᡸ"], ["ᢀ", "ᢄ"], ["ᢇ", "ᢨ"], "ᢪ", ["ᢰ", "ᣵ"], ["ᤀ", "ᤞ"], ["ᥐ", "ᥭ"], ["ᥰ", "ᥴ"], ["ᦀ", "ᦫ"], ["ᦰ", "ᧉ"], ["ᨀ", "ᨖ"], ["ᨠ", "ᩔ"], ["ᬅ", "ᬳ"], ["ᭅ", "ᭋ"], ["ᮃ", "ᮠ"], ["ᮮ", "ᮯ"], ["ᮺ", "ᯥ"], ["ᰀ", "ᰣ"], ["ᱍ", "ᱏ"], ["ᱚ", "ᱷ"], ["ᳩ", "ᳬ"], ["ᳮ", "ᳱ"], ["ᳵ", "ᳶ"], ["ℵ", "ℸ"], ["ⴰ", "ⵧ"], ["ⶀ", "ⶖ"], ["ⶠ", "ⶦ"], ["ⶨ", "ⶮ"], ["ⶰ", "ⶶ"], ["ⶸ", "ⶾ"], ["ⷀ", "ⷆ"], ["ⷈ", "ⷎ"], ["ⷐ", "ⷖ"], ["ⷘ", "ⷞ"], "〆", "〼", ["ぁ", "ゖ"], "ゟ", ["ァ", "ヺ"], "ヿ", ["ㄅ", "ㄯ"], ["ㄱ", "ㆎ"], ["ㆠ", "ㆺ"], ["ㇰ", "ㇿ"], ["㐀", "䶵"], ["一", "鿯"], ["ꀀ", "ꀔ"], ["ꀖ", "ꒌ"], ["ꓐ", "ꓷ"], ["ꔀ", "ꘋ"], ["ꘐ", "ꘟ"], ["ꘪ", "ꘫ"], "ꙮ", ["ꚠ", "ꛥ"], "ꞏ", "ꟷ", ["ꟻ", "ꠁ"], ["ꠃ", "ꠅ"], ["ꠇ", "ꠊ"], ["ꠌ", "ꠢ"], ["ꡀ", "ꡳ"], ["ꢂ", "ꢳ"], ["ꣲ", "ꣷ"], "ꣻ", ["ꣽ", "ꣾ"], ["ꤊ", "ꤥ"], ["ꤰ", "ꥆ"], ["ꥠ", "ꥼ"], ["ꦄ", "ꦲ"], ["ꧠ", "ꧤ"], ["ꧧ", "ꧯ"], ["ꧺ", "ꧾ"], ["ꨀ", "ꨨ"], ["ꩀ", "ꩂ"], ["ꩄ", "ꩋ"], ["ꩠ", "ꩯ"], ["ꩱ", "ꩶ"], "ꩺ", ["ꩾ", "ꪯ"], "ꪱ", ["ꪵ", "ꪶ"], ["ꪹ", "ꪽ"], "ꫀ", "ꫂ", ["ꫛ", "ꫜ"], ["ꫠ", "ꫪ"], "ꫲ", ["ꬁ", "ꬆ"], ["ꬉ", "ꬎ"], ["ꬑ", "ꬖ"], ["ꬠ", "ꬦ"], ["ꬨ", "ꬮ"], ["ꯀ", "ꯢ"], ["가", "힣"], ["ힰ", "ퟆ"], ["ퟋ", "ퟻ"], ["豈", "舘"], ["並", "龎"], "יִ", ["ײַ", "ﬨ"], ["שׁ", "זּ"], ["טּ", "לּ"], "מּ", ["נּ", "סּ"], ["ףּ", "פּ"], ["צּ", "ﮱ"], ["ﯓ", "ﴽ"], ["ﵐ", "ﶏ"], ["ﶒ", "ﷇ"], ["ﷰ", "ﷻ"], ["ﹰ", "ﹴ"], ["ﹶ", "ﻼ"], ["ｦ", "ｯ"], ["ｱ", "ﾝ"], ["ﾠ", "ﾾ"], ["ￂ", "ￇ"], ["ￊ", "ￏ"], ["ￒ", "ￗ"], ["ￚ", "ￜ"]], !1, !1), s1 = y(["ǅ", "ǈ", "ǋ", "ǲ", ["ᾈ", "ᾏ"], ["ᾘ", "ᾟ"], ["ᾨ", "ᾯ"], "ᾼ", "ῌ", "ῼ"], !1, !1), C1 = y([["A", "Z"], ["À", "Ö"], ["Ø", "Þ"], "Ā", "Ă", "Ą", "Ć", "Ĉ", "Ċ", "Č", "Ď", "Đ", "Ē", "Ĕ", "Ė", "Ę", "Ě", "Ĝ", "Ğ", "Ġ", "Ģ", "Ĥ", "Ħ", "Ĩ", "Ī", "Ĭ", "Į", "İ", "Ĳ", "Ĵ", "Ķ", "Ĺ", "Ļ", "Ľ", "Ŀ", "Ł", "Ń", "Ņ", "Ň", "Ŋ", "Ō", "Ŏ", "Ő", "Œ", "Ŕ", "Ŗ", "Ř", "Ś", "Ŝ", "Ş", "Š", "Ţ", "Ť", "Ŧ", "Ũ", "Ū", "Ŭ", "Ů", "Ű", "Ų", "Ŵ", "Ŷ", ["Ÿ", "Ź"], "Ż", "Ž", ["Ɓ", "Ƃ"], "Ƅ", ["Ɔ", "Ƈ"], ["Ɖ", "Ƌ"], ["Ǝ", "Ƒ"], ["Ɠ", "Ɣ"], ["Ɩ", "Ƙ"], ["Ɯ", "Ɲ"], ["Ɵ", "Ơ"], "Ƣ", "Ƥ", ["Ʀ", "Ƨ"], "Ʃ", "Ƭ", ["Ʈ", "Ư"], ["Ʊ", "Ƴ"], "Ƶ", ["Ʒ", "Ƹ"], "Ƽ", "Ǆ", "Ǉ", "Ǌ", "Ǎ", "Ǐ", "Ǒ", "Ǔ", "Ǖ", "Ǘ", "Ǚ", "Ǜ", "Ǟ", "Ǡ", "Ǣ", "Ǥ", "Ǧ", "Ǩ", "Ǫ", "Ǭ", "Ǯ", "Ǳ", "Ǵ", ["Ƕ", "Ǹ"], "Ǻ", "Ǽ", "Ǿ", "Ȁ", "Ȃ", "Ȅ", "Ȇ", "Ȉ", "Ȋ", "Ȍ", "Ȏ", "Ȑ", "Ȓ", "Ȕ", "Ȗ", "Ș", "Ț", "Ȝ", "Ȟ", "Ƞ", "Ȣ", "Ȥ", "Ȧ", "Ȩ", "Ȫ", "Ȭ", "Ȯ", "Ȱ", "Ȳ", ["Ⱥ", "Ȼ"], ["Ƚ", "Ⱦ"], "Ɂ", ["Ƀ", "Ɇ"], "Ɉ", "Ɋ", "Ɍ", "Ɏ", "Ͱ", "Ͳ", "Ͷ", "Ϳ", "Ά", ["Έ", "Ί"], "Ό", ["Ύ", "Ώ"], ["Α", "Ρ"], ["Σ", "Ϋ"], "Ϗ", ["ϒ", "ϔ"], "Ϙ", "Ϛ", "Ϝ", "Ϟ", "Ϡ", "Ϣ", "Ϥ", "Ϧ", "Ϩ", "Ϫ", "Ϭ", "Ϯ", "ϴ", "Ϸ", ["Ϲ", "Ϻ"], ["Ͻ", "Я"], "Ѡ", "Ѣ", "Ѥ", "Ѧ", "Ѩ", "Ѫ", "Ѭ", "Ѯ", "Ѱ", "Ѳ", "Ѵ", "Ѷ", "Ѹ", "Ѻ", "Ѽ", "Ѿ", "Ҁ", "Ҋ", "Ҍ", "Ҏ", "Ґ", "Ғ", "Ҕ", "Җ", "Ҙ", "Қ", "Ҝ", "Ҟ", "Ҡ", "Ң", "Ҥ", "Ҧ", "Ҩ", "Ҫ", "Ҭ", "Ү", "Ұ", "Ҳ", "Ҵ", "Ҷ", "Ҹ", "Һ", "Ҽ", "Ҿ", ["Ӏ", "Ӂ"], "Ӄ", "Ӆ", "Ӈ", "Ӊ", "Ӌ", "Ӎ", "Ӑ", "Ӓ", "Ӕ", "Ӗ", "Ә", "Ӛ", "Ӝ", "Ӟ", "Ӡ", "Ӣ", "Ӥ", "Ӧ", "Ө", "Ӫ", "Ӭ", "Ӯ", "Ӱ", "Ӳ", "Ӵ", "Ӷ", "Ӹ", "Ӻ", "Ӽ", "Ӿ", "Ԁ", "Ԃ", "Ԅ", "Ԇ", "Ԉ", "Ԋ", "Ԍ", "Ԏ", "Ԑ", "Ԓ", "Ԕ", "Ԗ", "Ԙ", "Ԛ", "Ԝ", "Ԟ", "Ԡ", "Ԣ", "Ԥ", "Ԧ", "Ԩ", "Ԫ", "Ԭ", "Ԯ", ["Ա", "Ֆ"], ["Ⴀ", "Ⴥ"], "Ⴧ", "Ⴭ", ["Ꭰ", "Ᏽ"], ["Ა", "Ჺ"], ["Ჽ", "Ჿ"], "Ḁ", "Ḃ", "Ḅ", "Ḇ", "Ḉ", "Ḋ", "Ḍ", "Ḏ", "Ḑ", "Ḓ", "Ḕ", "Ḗ", "Ḙ", "Ḛ", "Ḝ", "Ḟ", "Ḡ", "Ḣ", "Ḥ", "Ḧ", "Ḩ", "Ḫ", "Ḭ", "Ḯ", "Ḱ", "Ḳ", "Ḵ", "Ḷ", "Ḹ", "Ḻ", "Ḽ", "Ḿ", "Ṁ", "Ṃ", "Ṅ", "Ṇ", "Ṉ", "Ṋ", "Ṍ", "Ṏ", "Ṑ", "Ṓ", "Ṕ", "Ṗ", "Ṙ", "Ṛ", "Ṝ", "Ṟ", "Ṡ", "Ṣ", "Ṥ", "Ṧ", "Ṩ", "Ṫ", "Ṭ", "Ṯ", "Ṱ", "Ṳ", "Ṵ", "Ṷ", "Ṹ", "Ṻ", "Ṽ", "Ṿ", "Ẁ", "Ẃ", "Ẅ", "Ẇ", "Ẉ", "Ẋ", "Ẍ", "Ẏ", "Ẑ", "Ẓ", "Ẕ", "ẞ", "Ạ", "Ả", "Ấ", "Ầ", "Ẩ", "Ẫ", "Ậ", "Ắ", "Ằ", "Ẳ", "Ẵ", "Ặ", "Ẹ", "Ẻ", "Ẽ", "Ế", "Ề", "Ể", "Ễ", "Ệ", "Ỉ", "Ị", "Ọ", "Ỏ", "Ố", "Ồ", "Ổ", "Ỗ", "Ộ", "Ớ", "Ờ", "Ở", "Ỡ", "Ợ", "Ụ", "Ủ", "Ứ", "Ừ", "Ử", "Ữ", "Ự", "Ỳ", "Ỵ", "Ỷ", "Ỹ", "Ỻ", "Ỽ", "Ỿ", ["Ἀ", "Ἇ"], ["Ἐ", "Ἕ"], ["Ἠ", "Ἧ"], ["Ἰ", "Ἷ"], ["Ὀ", "Ὅ"], "Ὑ", "Ὓ", "Ὕ", "Ὗ", ["Ὠ", "Ὧ"], ["Ᾰ", "Ά"], ["Ὲ", "Ή"], ["Ῐ", "Ί"], ["Ῠ", "Ῥ"], ["Ὸ", "Ώ"], "ℂ", "ℇ", ["ℋ", "ℍ"], ["ℐ", "ℒ"], "ℕ", ["ℙ", "ℝ"], "ℤ", "Ω", "ℨ", ["K", "ℭ"], ["ℰ", "ℳ"], ["ℾ", "ℿ"], "ⅅ", "Ↄ", ["Ⰰ", "Ⱞ"], "Ⱡ", ["Ɫ", "Ɽ"], "Ⱨ", "Ⱪ", "Ⱬ", ["Ɑ", "Ɒ"], "Ⱳ", "Ⱶ", ["Ȿ", "Ⲁ"], "Ⲃ", "Ⲅ", "Ⲇ", "Ⲉ", "Ⲋ", "Ⲍ", "Ⲏ", "Ⲑ", "Ⲓ", "Ⲕ", "Ⲗ", "Ⲙ", "Ⲛ", "Ⲝ", "Ⲟ", "Ⲡ", "Ⲣ", "Ⲥ", "Ⲧ", "Ⲩ", "Ⲫ", "Ⲭ", "Ⲯ", "Ⲱ", "Ⲳ", "Ⲵ", "Ⲷ", "Ⲹ", "Ⲻ", "Ⲽ", "Ⲿ", "Ⳁ", "Ⳃ", "Ⳅ", "Ⳇ", "Ⳉ", "Ⳋ", "Ⳍ", "Ⳏ", "Ⳑ", "Ⳓ", "Ⳕ", "Ⳗ", "Ⳙ", "Ⳛ", "Ⳝ", "Ⳟ", "Ⳡ", "Ⳣ", "Ⳬ", "Ⳮ", "Ⳳ", "Ꙁ", "Ꙃ", "Ꙅ", "Ꙇ", "Ꙉ", "Ꙋ", "Ꙍ", "Ꙏ", "Ꙑ", "Ꙓ", "Ꙕ", "Ꙗ", "Ꙙ", "Ꙛ", "Ꙝ", "Ꙟ", "Ꙡ", "Ꙣ", "Ꙥ", "Ꙧ", "Ꙩ", "Ꙫ", "Ꙭ", "Ꚁ", "Ꚃ", "Ꚅ", "Ꚇ", "Ꚉ", "Ꚋ", "Ꚍ", "Ꚏ", "Ꚑ", "Ꚓ", "Ꚕ", "Ꚗ", "Ꚙ", "Ꚛ", "Ꜣ", "Ꜥ", "Ꜧ", "Ꜩ", "Ꜫ", "Ꜭ", "Ꜯ", "Ꜳ", "Ꜵ", "Ꜷ", "Ꜹ", "Ꜻ", "Ꜽ", "Ꜿ", "Ꝁ", "Ꝃ", "Ꝅ", "Ꝇ", "Ꝉ", "Ꝋ", "Ꝍ", "Ꝏ", "Ꝑ", "Ꝓ", "Ꝕ", "Ꝗ", "Ꝙ", "Ꝛ", "Ꝝ", "Ꝟ", "Ꝡ", "Ꝣ", "Ꝥ", "Ꝧ", "Ꝩ", "Ꝫ", "Ꝭ", "Ꝯ", "Ꝺ", "Ꝼ", ["Ᵹ", "Ꝿ"], "Ꞁ", "Ꞃ", "Ꞅ", "Ꞇ", "Ꞌ", "Ɥ", "Ꞑ", "Ꞓ", "Ꞗ", "Ꞙ", "Ꞛ", "Ꞝ", "Ꞟ", "Ꞡ", "Ꞣ", "Ꞥ", "Ꞧ", "Ꞩ", ["Ɦ", "Ɪ"], ["Ʞ", "Ꞵ"], "Ꞷ", "Ꞹ", ["Ａ", "Ｚ"]], !1, !1), F1 = y([["ᛮ", "ᛰ"], ["Ⅰ", "ↂ"], ["ↅ", "ↈ"], "〇", ["〡", "〩"], ["〸", "〺"], ["ꛦ", "ꛯ"]], !1, !1), t1 = function() {
    return parseFloat(X1());
  }, n1 = function(u) {
    const A = u.join("");
    return A === "." || A === "_";
  }, a1 = function(u) {
    return new We(u.join(""));
  }, f1 = function(u) {
    return u;
  }, i1 = function(u, A) {
    return u.arguments_.stepsPerCycle = A, u;
  }, B1 = function(u) {
    return u;
  }, D1 = function(u) {
    return u.arguments_.alignment = "polymeter_slowcat", u;
  }, o1 = function(u) {
    return (A) => A.options_.weight = (A.options_.weight ?? 1) + (u ?? 2) - 1;
  }, c1 = function(u) {
    return (A) => A.options_.reps = (A.options_.reps ?? 1) + (u ?? 2) - 1;
  }, l1 = function(u, A, E) {
    return (C) => C.options_.ops.push({ type_: "bjorklund", arguments_: { pulse: u, step: A, rotation: E } });
  }, p1 = function(u) {
    return (A) => A.options_.ops.push({ type_: "stretch", arguments_: { amount: u, type: "slow" } });
  }, g1 = function(u) {
    return (A) => A.options_.ops.push({ type_: "stretch", arguments_: { amount: u, type: "fast" } });
  }, v1 = function(u) {
    return (A) => A.options_.ops.push({ type_: "degradeBy", arguments_: { amount: u, seed: eu++ } });
  }, $1 = function(u) {
    return (A) => A.options_.ops.push({ type_: "tail", arguments_: { element: u } });
  }, h1 = function(u) {
    return (A) => A.options_.ops.push({ type_: "range", arguments_: { element: u } });
  }, m1 = function(u, A) {
    const E = new Ge(u, { ops: [], weight: 1, reps: 1 });
    for (const C of A)
      C(E);
    return E;
  }, _1 = function(u) {
    return new X(u, "fastcat");
  }, d1 = function(u) {
    return { alignment: "stack", list: u };
  }, y1 = function(u) {
    return { alignment: "rand", list: u, seed: eu++ };
  }, w1 = function(u) {
    return { alignment: "feet", list: u, seed: eu++ };
  }, x1 = function(u, A) {
    return A && A.list.length > 0 ? new X([u, ...A.list], A.alignment, A.seed) : u;
  }, b1 = function(u, A) {
    return new X(A ? [u, ...A.list] : [u], "polymeter");
  }, j1 = function(u) {
    return u;
  }, k1 = function(u) {
    return { name: "struct", args: { mini: u } };
  }, S1 = function(u) {
    return { name: "target", args: { name: u } };
  }, L1 = function(u, A, E) {
    return { name: "bjorklund", args: { pulse: u, step: parseInt(A) } };
  }, P1 = function(u) {
    return { name: "stretch", args: { amount: u } };
  }, R1 = function(u) {
    return { name: "shift", args: { amount: "-" + u } };
  }, q1 = function(u) {
    return { name: "shift", args: { amount: u } };
  }, I1 = function(u) {
    return { name: "stretch", args: { amount: "1/" + u } };
  }, N1 = function(u) {
    return { name: "scale", args: { scale: u.join("") } };
  }, xu = function(u, A) {
    return A;
  }, O1 = function(u, A) {
    return A.unshift(u), new X(A, "slowcat");
  }, T1 = function(u) {
    return u;
  }, z1 = function(u, A) {
    return new Ze(u.name, u.args, A);
  }, M1 = function(u) {
    return u;
  }, W1 = function(u) {
    return u;
  }, Z1 = function(u) {
    return new uu("setcps", { value: u });
  }, G1 = function(u) {
    return new uu("setcps", { value: u / 120 / 2 });
  }, U1 = function() {
    return new uu("hush");
  }, r = 0, D = 0, W = [{ line: 1, column: 1 }], x = 0, K = [], F = 0, Z;
  if ("startRule" in n) {
    if (!(n.startRule in l))
      throw new Error(`Can't start parsing from rule "` + n.startRule + '".');
    p = l[n.startRule];
  }
  function X1() {
    return s.substring(D, r);
  }
  function bu() {
    return V(D, r);
  }
  function o(u, A) {
    return { type: "literal", text: u, ignoreCase: A };
  }
  function y(u, A, E) {
    return { type: "class", parts: u, inverted: A, ignoreCase: E };
  }
  function H1() {
    return { type: "end" };
  }
  function Q(u) {
    return { type: "other", description: u };
  }
  function ju(u) {
    var A = W[u], E;
    if (A)
      return A;
    for (E = u - 1; !W[E]; )
      E--;
    for (A = W[E], A = {
      line: A.line,
      column: A.column
    }; E < u; )
      s.charCodeAt(E) === 10 ? (A.line++, A.column = 1) : A.column++, E++;
    return W[u] = A, A;
  }
  function V(u, A, E) {
    var C = ju(u), a = ju(A), v = {
      source: c,
      start: {
        offset: u,
        line: C.line,
        column: C.column
      },
      end: {
        offset: A,
        line: a.line,
        column: a.column
      }
    };
    return E && c && typeof c.offset == "function" && (v.start = c.offset(v.start), v.end = c.offset(v.end)), v;
  }
  function t(u) {
    r < x || (r > x && (x = r, K = []), K.push(u));
  }
  function J1(u, A, E) {
    return new N(
      N.buildMessage(u, A),
      u,
      A,
      E
    );
  }
  function ku() {
    var u;
    return u = Re(), u;
  }
  function b() {
    var u, A;
    return F++, u = r, Su(), A = G(), A !== e ? (ue(), Y1(), D = u, u = t1()) : (r = u, u = e), F--, u === e && F === 0 && t(B0), u;
  }
  function K1() {
    var u;
    return s.charCodeAt(r) === 46 ? (u = m, r++) : (u = e, F === 0 && t(J)), u;
  }
  function Q1() {
    var u;
    return r0.test(s.charAt(r)) ? (u = s.charAt(r), r++) : (u = e, F === 0 && t(D0)), u;
  }
  function V1() {
    var u;
    return E0.test(s.charAt(r)) ? (u = s.charAt(r), r++) : (u = e, F === 0 && t(o0)), u;
  }
  function Y1() {
    var u, A, E, C, a;
    if (u = r, A = V1(), A !== e) {
      if (E = Su(), E === e && (E = ee()), E === e && (E = null), C = [], a = S(), a !== e)
        for (; a !== e; )
          C.push(a), a = S();
      else
        C = e;
      C !== e ? (A = [A, E, C], u = A) : (r = u, u = e);
    } else
      r = u, u = e;
    return u;
  }
  function ue() {
    var u, A, E, C;
    if (u = r, A = K1(), A !== e) {
      if (E = [], C = S(), C !== e)
        for (; C !== e; )
          E.push(C), C = S();
      else
        E = e;
      E !== e ? (A = [A, E], u = A) : (r = u, u = e);
    } else
      r = u, u = e;
    return u;
  }
  function G() {
    var u, A, E, C;
    if (u = Ae(), u === e)
      if (u = r, A = Q1(), A !== e) {
        for (E = [], C = S(); C !== e; )
          E.push(C), C = S();
        A = [A, E], u = A;
      } else
        r = u, u = e;
    return u;
  }
  function Su() {
    var u;
    return s.charCodeAt(r) === 45 ? (u = $, r++) : (u = e, F === 0 && t(hu)), u;
  }
  function ee() {
    var u;
    return s.charCodeAt(r) === 43 ? (u = _, r++) : (u = e, F === 0 && t(c0)), u;
  }
  function Ae() {
    var u;
    return s.charCodeAt(r) === 48 ? (u = i, r++) : (u = e, F === 0 && t(l0)), u;
  }
  function S() {
    var u;
    return s0.test(s.charAt(r)) ? (u = s.charAt(r), r++) : (u = e, F === 0 && t(p0)), u;
  }
  function f() {
    var u, A;
    for (F++, u = [], vu.test(s.charAt(r)) ? (A = s.charAt(r), r++) : (A = e, F === 0 && t(mu)); A !== e; )
      u.push(A), vu.test(s.charAt(r)) ? (A = s.charAt(r), r++) : (A = e, F === 0 && t(mu));
    return F--, A = e, F === 0 && t(g0), u;
  }
  function L() {
    var u, A, E, C;
    return u = r, A = f(), s.charCodeAt(r) === 44 ? (E = B, r++) : (E = e, F === 0 && t(v0)), E !== e ? (C = f(), A = [A, E, C], u = A) : (r = u, u = e), u;
  }
  function Lu() {
    var u, A, E, C;
    return u = r, A = f(), s.charCodeAt(r) === 124 ? (E = g, r++) : (E = e, F === 0 && t($0)), E !== e ? (C = f(), A = [A, E, C], u = A) : (r = u, u = e), u;
  }
  function Pu() {
    var u, A, E, C;
    return u = r, A = f(), s.charCodeAt(r) === 46 ? (E = m, r++) : (E = e, F === 0 && t(J)), E !== e ? (C = f(), A = [A, E, C], u = A) : (r = u, u = e), u;
  }
  function P() {
    var u;
    return s.charCodeAt(r) === 34 ? (u = d, r++) : (u = e, F === 0 && t(h0)), u === e && (s.charCodeAt(r) === 39 ? (u = k, r++) : (u = e, F === 0 && t(m0))), u;
  }
  function U() {
    var u;
    return F++, u = qe(), u === e && (C0.test(s.charAt(r)) ? (u = s.charAt(r), r++) : (u = e, F === 0 && t(d0)), u === e && (s.charCodeAt(r) === 45 ? (u = $, r++) : (u = e, F === 0 && t(hu)), u === e && (s.charCodeAt(r) === 35 ? (u = z, r++) : (u = e, F === 0 && t(y0)), u === e && (s.charCodeAt(r) === 46 ? (u = m, r++) : (u = e, F === 0 && t(J)), u === e && (s.charCodeAt(r) === 94 ? (u = M, r++) : (u = e, F === 0 && t(w0)), u === e && (s.charCodeAt(r) === 95 ? (u = ru, r++) : (u = e, F === 0 && t(_u)))))))), F--, u === e && F === 0 && t(_0), u;
  }
  function Ru() {
    var u, A, E, C;
    if (u = r, f(), A = [], E = U(), E !== e)
      for (; E !== e; )
        A.push(E), E = U();
    else
      A = e;
    return A !== e ? (E = f(), D = r, C = n1(A), C ? C = e : C = void 0, C !== e ? (D = u, u = a1(A)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function re() {
    var u, A, E, C;
    return u = r, f(), s.charCodeAt(r) === 91 ? (A = Eu, r++) : (A = e, F === 0 && t(du)), A !== e ? (f(), E = Nu(), E !== e ? (f(), s.charCodeAt(r) === 93 ? (C = su, r++) : (C = e, F === 0 && t(yu)), C !== e ? (f(), D = u, u = f1(E)) : (r = u, u = e)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function Ee() {
    var u, A, E, C, a;
    return u = r, f(), s.charCodeAt(r) === 123 ? (A = Wu, r++) : (A = e, F === 0 && t(x0)), A !== e ? (f(), E = Ou(), E !== e ? (f(), s.charCodeAt(r) === 125 ? (C = Zu, r++) : (C = e, F === 0 && t(b0)), C !== e ? (a = se(), a === e && (a = null), f(), D = u, u = i1(E, a)) : (r = u, u = e)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function se() {
    var u, A, E;
    return u = r, s.charCodeAt(r) === 37 ? (A = Gu, r++) : (A = e, F === 0 && t(j0)), A !== e ? (E = R(), E !== e ? (D = u, u = B1(E)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function Ce() {
    var u, A, E, C;
    return u = r, f(), s.charCodeAt(r) === 60 ? (A = Uu, r++) : (A = e, F === 0 && t(k0)), A !== e ? (f(), E = Ou(), E !== e ? (f(), s.charCodeAt(r) === 62 ? (C = Xu, r++) : (C = e, F === 0 && t(S0)), C !== e ? (f(), D = u, u = D1(E)) : (r = u, u = e)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function R() {
    var u;
    return u = Ru(), u === e && (u = re(), u === e && (u = Ee(), u === e && (u = Ce()))), u;
  }
  function qu() {
    var u;
    return u = Fe(), u === e && (u = ne(), u === e && (u = ae(), u === e && (u = fe(), u === e && (u = te(), u === e && (u = ie(), u === e && (u = Be(), u === e && (u = De()))))))), u;
  }
  function Fe() {
    var u, A, E;
    return u = r, f(), s.charCodeAt(r) === 64 ? (A = Hu, r++) : (A = e, F === 0 && t(L0)), A === e && (s.charCodeAt(r) === 95 ? (A = ru, r++) : (A = e, F === 0 && t(_u))), A !== e ? (E = b(), E === e && (E = null), D = u, u = o1(E)) : (r = u, u = e), u;
  }
  function te() {
    var u, A, E;
    return u = r, f(), s.charCodeAt(r) === 33 ? (A = Ju, r++) : (A = e, F === 0 && t(P0)), A !== e ? (E = b(), E === e && (E = null), D = u, u = c1(E)) : (r = u, u = e), u;
  }
  function ne() {
    var u, A, E, C, a, v, w;
    return u = r, s.charCodeAt(r) === 40 ? (A = Ku, r++) : (A = e, F === 0 && t(R0)), A !== e ? (f(), E = O(), E !== e ? (f(), C = L(), C !== e ? (f(), a = O(), a !== e ? (f(), L(), f(), v = O(), v === e && (v = null), f(), s.charCodeAt(r) === 41 ? (w = Qu, r++) : (w = e, F === 0 && t(q0)), w !== e ? (D = u, u = l1(E, a, v)) : (r = u, u = e)) : (r = u, u = e)) : (r = u, u = e)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function ae() {
    var u, A, E;
    return u = r, s.charCodeAt(r) === 47 ? (A = Vu, r++) : (A = e, F === 0 && t(I0)), A !== e ? (E = R(), E !== e ? (D = u, u = p1(E)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function fe() {
    var u, A, E;
    return u = r, s.charCodeAt(r) === 42 ? (A = Yu, r++) : (A = e, F === 0 && t(N0)), A !== e ? (E = R(), E !== e ? (D = u, u = g1(E)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function ie() {
    var u, A, E;
    return u = r, s.charCodeAt(r) === 63 ? (A = u0, r++) : (A = e, F === 0 && t(O0)), A !== e ? (E = b(), E === e && (E = null), D = u, u = v1(E)) : (r = u, u = e), u;
  }
  function Be() {
    var u, A, E;
    return u = r, s.charCodeAt(r) === 58 ? (A = e0, r++) : (A = e, F === 0 && t(T0)), A !== e ? (E = R(), E !== e ? (D = u, u = $1(E)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function De() {
    var u, A, E;
    return u = r, s.substr(r, 2) === Cu ? (A = Cu, r += 2) : (A = e, F === 0 && t(z0)), A !== e ? (E = R(), E !== e ? (D = u, u = h1(E)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function O() {
    var u, A, E, C;
    if (u = r, A = R(), A !== e) {
      for (E = [], C = qu(); C !== e; )
        E.push(C), C = qu();
      D = u, u = m1(A, E);
    } else
      r = u, u = e;
    return u;
  }
  function j() {
    var u, A, E;
    if (u = r, A = [], E = O(), E !== e)
      for (; E !== e; )
        A.push(E), E = O();
    else
      A = e;
    return A !== e && (D = u, A = _1(A)), u = A, u;
  }
  function Iu() {
    var u, A, E, C, a;
    if (u = r, A = [], E = r, C = L(), C !== e ? (a = j(), a !== e ? E = a : (r = E, E = e)) : (r = E, E = e), E !== e)
      for (; E !== e; )
        A.push(E), E = r, C = L(), C !== e ? (a = j(), a !== e ? E = a : (r = E, E = e)) : (r = E, E = e);
    else
      A = e;
    return A !== e && (D = u, A = d1(A)), u = A, u;
  }
  function oe() {
    var u, A, E, C, a;
    if (u = r, A = [], E = r, C = Lu(), C !== e ? (a = j(), a !== e ? E = a : (r = E, E = e)) : (r = E, E = e), E !== e)
      for (; E !== e; )
        A.push(E), E = r, C = Lu(), C !== e ? (a = j(), a !== e ? E = a : (r = E, E = e)) : (r = E, E = e);
    else
      A = e;
    return A !== e && (D = u, A = y1(A)), u = A, u;
  }
  function ce() {
    var u, A, E, C, a;
    if (u = r, A = [], E = r, C = Pu(), C !== e ? (a = j(), a !== e ? E = a : (r = E, E = e)) : (r = E, E = e), E !== e)
      for (; E !== e; )
        A.push(E), E = r, C = Pu(), C !== e ? (a = j(), a !== e ? E = a : (r = E, E = e)) : (r = E, E = e);
    else
      A = e;
    return A !== e && (D = u, A = w1(A)), u = A, u;
  }
  function Nu() {
    var u, A, E;
    return u = r, A = j(), A !== e ? (E = Iu(), E === e && (E = oe(), E === e && (E = ce())), E === e && (E = null), D = u, u = x1(A, E)) : (r = u, u = e), u;
  }
  function Ou() {
    var u, A, E;
    return u = r, A = j(), A !== e ? (E = Iu(), E === e && (E = null), D = u, u = b1(A, E)) : (r = u, u = e), u;
  }
  function le() {
    var u, A, E, C;
    return u = r, f(), A = P(), A !== e ? (f(), E = Nu(), E !== e ? (f(), C = P(), C !== e ? (D = u, u = j1(E)) : (r = u, u = e)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function pe() {
    var u;
    return u = ye(), u === e && (u = he(), u === e && (u = de(), u === e && (u = ve(), u === e && (u = $e(), u === e && (u = ge(), u === e && (u = _e(), u === e && (u = me()))))))), u;
  }
  function ge() {
    var u, A, E;
    return u = r, s.substr(r, 6) === Fu ? (A = Fu, r += 6) : (A = e, F === 0 && t(M0)), A !== e ? (f(), E = q(), E !== e ? (D = u, u = k1(E)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function ve() {
    var u, A, E, C, a;
    return u = r, s.substr(r, 6) === tu ? (A = tu, r += 6) : (A = e, F === 0 && t(W0)), A !== e ? (f(), E = P(), E !== e ? (C = Ru(), C !== e ? (a = P(), a !== e ? (D = u, u = S1(C)) : (r = u, u = e)) : (r = u, u = e)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function $e() {
    var u, A, E, C;
    return u = r, s.substr(r, 6) === nu ? (A = nu, r += 6) : (A = e, F === 0 && t(Z0)), A !== e ? (f(), E = G(), E !== e ? (f(), C = G(), C !== e ? (f(), G(), D = u, u = L1(E, C)) : (r = u, u = e)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function he() {
    var u, A, E;
    return u = r, s.substr(r, 4) === au ? (A = au, r += 4) : (A = e, F === 0 && t(G0)), A !== e ? (f(), E = b(), E !== e ? (D = u, u = P1(E)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function me() {
    var u, A, E;
    return u = r, s.substr(r, 4) === fu ? (A = fu, r += 4) : (A = e, F === 0 && t(U0)), A !== e ? (f(), E = b(), E !== e ? (D = u, u = R1(E)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function _e() {
    var u, A, E;
    return u = r, s.substr(r, 4) === iu ? (A = iu, r += 4) : (A = e, F === 0 && t(X0)), A !== e ? (f(), E = b(), E !== e ? (D = u, u = q1(E)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function de() {
    var u, A, E;
    return u = r, s.substr(r, 4) === Bu ? (A = Bu, r += 4) : (A = e, F === 0 && t(H0)), A !== e ? (f(), E = b(), E !== e ? (D = u, u = I1(E)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function ye() {
    var u, A, E, C, a;
    if (u = r, s.substr(r, 5) === Du ? (A = Du, r += 5) : (A = e, F === 0 && t(J0)), A !== e)
      if (f(), E = P(), E !== e) {
        if (C = [], a = U(), a !== e)
          for (; a !== e; )
            C.push(a), a = U();
        else
          C = e;
        C !== e ? (a = P(), a !== e ? (D = u, u = N1(C)) : (r = u, u = e)) : (r = u, u = e);
      } else
        r = u, u = e;
    else
      r = u, u = e;
    return u;
  }
  function Y() {
    var u, A, E, C;
    if (u = r, s.substr(r, 2) === ou ? (A = ou, r += 2) : (A = e, F === 0 && t(K0)), A !== e) {
      for (E = [], $u.test(s.charAt(r)) ? (C = s.charAt(r), r++) : (C = e, F === 0 && t(wu)); C !== e; )
        E.push(C), $u.test(s.charAt(r)) ? (C = s.charAt(r), r++) : (C = e, F === 0 && t(wu));
      A = [A, E], u = A;
    } else
      r = u, u = e;
    return u;
  }
  function we() {
    var u, A, E, C, a, v, w, I;
    if (u = r, s.substr(r, 3) === cu ? (A = cu, r += 3) : (A = e, F === 0 && t(Q0)), A !== e)
      if (f(), s.charCodeAt(r) === 91 ? (E = Eu, r++) : (E = e, F === 0 && t(du)), E !== e)
        if (f(), C = q(), C !== e) {
          for (a = [], v = r, w = L(), w !== e ? (I = q(), I !== e ? (D = v, v = xu(C, I)) : (r = v, v = e)) : (r = v, v = e); v !== e; )
            a.push(v), v = r, w = L(), w !== e ? (I = q(), I !== e ? (D = v, v = xu(C, I)) : (r = v, v = e)) : (r = v, v = e);
          v = f(), s.charCodeAt(r) === 93 ? (w = su, r++) : (w = e, F === 0 && t(yu)), w !== e ? (D = u, u = O1(C, a)) : (r = u, u = e);
        } else
          r = u, u = e;
      else
        r = u, u = e;
    else
      r = u, u = e;
    return u;
  }
  function xe() {
    var u;
    return u = we(), u === e && (u = le()), u;
  }
  function q() {
    var u, A, E, C, a;
    if (u = r, A = xe(), A !== e) {
      for (f(), E = [], C = Y(); C !== e; )
        E.push(C), C = Y();
      D = u, u = T1(A);
    } else
      r = u, u = e;
    return u === e && (u = r, A = pe(), A !== e ? (f(), s.charCodeAt(r) === 36 ? (E = A0, r++) : (E = e, F === 0 && t(V0)), E !== e ? (C = f(), a = q(), a !== e ? (D = u, u = z1(A, a)) : (r = u, u = e)) : (r = u, u = e)) : (r = u, u = e)), u;
  }
  function be() {
    var u, A;
    return u = r, A = q(), A !== e && (D = u, A = M1(A)), u = A, u === e && (u = Y()), u;
  }
  function je() {
    var u;
    return u = be(), u;
  }
  function ke() {
    var u, A;
    return u = r, f(), A = Se(), A === e && (A = Le(), A === e && (A = Pe())), A !== e ? (f(), D = u, u = W1(A)) : (r = u, u = e), u;
  }
  function Se() {
    var u, A, E;
    return u = r, s.substr(r, 6) === lu ? (A = lu, r += 6) : (A = e, F === 0 && t(Y0)), A !== e ? (f(), E = b(), E !== e ? (D = u, u = Z1(E)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function Le() {
    var u, A, E;
    return u = r, s.substr(r, 6) === pu ? (A = pu, r += 6) : (A = e, F === 0 && t(u1)), A !== e ? (f(), E = b(), E !== e ? (D = u, u = G1(E)) : (r = u, u = e)) : (r = u, u = e), u;
  }
  function Pe() {
    var u, A;
    return u = r, s.substr(r, 4) === gu ? (A = gu, r += 4) : (A = e, F === 0 && t(e1)), A !== e && (D = u, A = U1()), u = A, u;
  }
  function Re() {
    var u;
    return u = je(), u === e && (u = ke()), u;
  }
  function qe() {
    var u;
    return u = ze(), u === e && (u = Ie(), u === e && (u = Te(), u === e && (u = Ne(), u === e && (u = Oe(), u === e && (u = Me()))))), u;
  }
  function Ie() {
    var u;
    return F0.test(s.charAt(r)) ? (u = s.charAt(r), r++) : (u = e, F === 0 && t(A1)), u;
  }
  function Ne() {
    var u;
    return t0.test(s.charAt(r)) ? (u = s.charAt(r), r++) : (u = e, F === 0 && t(r1)), u;
  }
  function Oe() {
    var u;
    return n0.test(s.charAt(r)) ? (u = s.charAt(r), r++) : (u = e, F === 0 && t(E1)), u;
  }
  function Te() {
    var u;
    return a0.test(s.charAt(r)) ? (u = s.charAt(r), r++) : (u = e, F === 0 && t(s1)), u;
  }
  function ze() {
    var u;
    return f0.test(s.charAt(r)) ? (u = s.charAt(r), r++) : (u = e, F === 0 && t(C1)), u;
  }
  function Me() {
    var u;
    return i0.test(s.charAt(r)) ? (u = s.charAt(r), r++) : (u = e, F === 0 && t(F1)), u;
  }
  var We = function(u) {
    this.type_ = "atom", this.source_ = u, this.location_ = bu();
  }, X = function(u, A, E) {
    this.type_ = "pattern", this.arguments_ = { alignment: A }, E !== void 0 && (this.arguments_.seed = E), this.source_ = u;
  }, Ze = function(u, A, E) {
    this.type_ = u, this.arguments_ = A, this.source_ = E;
  }, Ge = function(u, A) {
    this.type_ = "element", this.source_ = u, this.options_ = A, this.location_ = bu();
  }, uu = function(u, A) {
    this.type_ = "command", this.name_ = u, this.options_ = A;
  }, eu = 0;
  if (Z = p(), Z !== e && r === s.length)
    return Z;
  throw Z !== e && r < s.length && t(H1()), J1(
    K,
    x < s.length ? s.charAt(x) : null,
    x < s.length ? V(x, x + 1) : V(x, x)
  );
}
const Tu = 3e-4, He = (s, n) => (e, c) => {
  const m = s.source_[c].options_?.ops;
  if (m)
    for (const $ of m)
      switch ($.type_) {
        case "stretch": {
          const _ = ["fast", "slow"], { type: i, amount: B } = $.arguments_;
          if (!_.includes(i))
            throw new Error(`mini: stretch: type must be one of ${_.join("|")} but got ${i}`);
          e = h.reify(e)[i](n(B));
          break;
        }
        case "bjorklund": {
          $.arguments_.rotation ? e = e.euclidRot(n($.arguments_.pulse), n($.arguments_.step), n($.arguments_.rotation)) : e = e.euclid(n($.arguments_.pulse), n($.arguments_.step));
          break;
        }
        case "degradeBy": {
          e = h.reify(e)._degradeByWith(h.rand.early(Tu * $.arguments_.seed), $.arguments_.amount ?? 0.5);
          break;
        }
        case "tail": {
          const _ = n($.arguments_.element);
          e = e.fmap((i) => (B) => Array.isArray(i) ? [...i, B] : [i, B]).appLeft(_);
          break;
        }
        case "range": {
          const _ = n($.arguments_.element);
          e = h.reify(e);
          const i = (g, d, k = 1) => Array.from(
            { length: Math.abs(d - g) / k + 1 },
            (z, M) => g < d ? g + M * k : g - M * k
          );
          e = ((g, d) => g.squeezeBind((k) => d.bind((z) => h.fastcat(...i(k, z)))))(e, _);
          break;
        }
        default:
          console.warn(`operator "${$.type_}" not implemented`);
      }
  return e;
};
function Je(s) {
  s.source_ = h.flatten(
    s.source_.map((n) => {
      const { reps: e } = n.options_ || {};
      return e ? (delete n.options_.reps, Array(e).fill(n)) : [n];
    })
  );
}
function T(s, n, e, c = 0) {
  e?.(s);
  const l = (p) => T(p, n, e, c);
  switch (s.type_) {
    case "pattern": {
      Je(s);
      const p = s.source_.map((i) => l(i)).map(He(s, l)), m = s.arguments_.alignment;
      if (m === "stack")
        return h.stack(...p);
      if (m === "polymeter_slowcat") {
        const i = p.map((B) => B._slow(h.Fraction(B.__weight ?? 1)));
        return h.stack(...i);
      }
      if (m === "polymeter") {
        const i = s.arguments_.stepsPerCycle ? l(s.arguments_.stepsPerCycle).fmap((g) => h.Fraction(g)) : h.pure(h.Fraction(p.length > 0 ? p[0].__weight : 1)), B = p.map((g) => g.fast(i.fmap((d) => d.div(g.__weight || 1))));
        return h.stack(...B);
      }
      if (m === "rand")
        return h.chooseInWith(h.rand.early(Tu * s.arguments_.seed).segment(1), p);
      if (m === "feet")
        return h.fastcat(...p);
      if (s.source_.some((i) => !!i.options_?.weight)) {
        const i = s.source_.reduce((g, d) => g + (d.options_?.weight || 1), 0), B = h.timeCat(...s.source_.map((g, d) => [g.options_?.weight || 1, p[d]]));
        return B.__weight = i, B;
      }
      const _ = h.sequence(...p);
      return _.__weight = p.length, _;
    }
    case "element":
      return l(s.source_);
    case "atom": {
      if (s.source_ === "~")
        return h.silence;
      if (!s.location_)
        return console.warn("no location for", s), s.source_;
      const p = isNaN(Number(s.source_)) ? s.source_ : Number(s.source_);
      if (c === -1)
        return h.pure(p);
      const [m, $] = zu(n, s, c);
      return h.pure(p).withLoc(m, $);
    }
    case "stretch":
      return l(s.source_).slow(l(s.arguments_.amount));
    default:
      return console.warn(`node type "${s.type_}" not implemented -> returning silence`), h.silence;
  }
}
const zu = (s, n, e = 0) => {
  const { start: c, end: l } = n.location_, p = s?.split("").slice(c.offset, l.offset).join(""), [m = 0, $ = 0] = p ? p.split(n.source_).map((_) => _.split("").filter((i) => i === " ").length) : [];
  return [c.offset + m + e, l.offset - $ + e];
}, H = (s, n, e) => {
  try {
    return Xe(s);
  } catch (c) {
    const l = [c.location.start.offset + n, c.location.end.offset + n], p = e.slice(0, l[0]).split(`
`).length;
    throw new Error(`[mini] parse error at line ${p}: ${c.message}`);
  }
}, Ke = (s, n, e) => {
  const c = H(s, n, e);
  let l = [];
  return T(
    c,
    s,
    (p) => {
      p.type_ === "atom" && l.push(p);
    },
    -1
  ), l;
}, Qe = (s, n = 0, e) => Ke(s, n, e).map((c) => zu(s, c, n)), Mu = (...s) => {
  const n = s.map((e) => {
    const c = `"${e}"`, l = H(c);
    return T(l, c);
  });
  return h.sequence(...n);
}, Ve = (s, n) => {
  const e = `"${s}"`, c = H(e);
  return T(c, e, null, n);
}, Ye = (s) => {
  const n = H(s);
  return T(n, s);
};
function uA(s) {
  return typeof s == "string" ? Mu(s) : h.reify(s);
}
function eA() {
  h.setStringParser(Mu);
}
export {
  N as SyntaxError,
  zu as getLeafLocation,
  Qe as getLeafLocations,
  Ke as getLeaves,
  Ye as h,
  Ve as m,
  Mu as mini,
  H as mini2ast,
  eA as miniAllStrings,
  uA as minify,
  Xe as parse,
  T as patternifyAST
};
